File: algorithms/23_elixir/sorting/insertion_sort.ex
defmodule InsertionSort do
  def sort(list), do: Enum.reduce(list, [], &insert/2)

  defp insert(x, []), do: [x]
  defp insert(x, [h | t]) when x <= h, do: [x, h | t]
  defp insert(x, [h | t]), do: [h | insert(x, t)]
end

data = [33, 7, 91, 12, 5, 5, 78, 2, 44, 19]
sorted = InsertionSort.sort(data)
unless sorted == Enum.sort(data), do: raise("not sorted")
IO.puts("[Elixir InsertionSort] Insertion sort verified: #{inspect(sorted)}")

File: algorithms/15_ruby/sorting/insertion_sort.rb
def insertion_sort(arr)
  a = arr.dup
  (1...a.length).each do |i|
    key = a[i]
    j = i - 1
    while j >= 0 && a[j] > key
      a[j + 1] = a[j]
      j -= 1
    end
    a[j + 1] = key
  end
  a
end

data = [33, 7, 91, 12, 5, 5, 78, 2, 44, 19]
sorted = insertion_sort(data)
raise "not sorted" unless sorted == data.sort
puts "[Ruby InsertionSort] Insertion sort verified: #{sorted.inspect}"
